metadata/src/main/java/org/apache/kafka/metadata/migration/KRaftMigrationZkWriter.java

```diff
@@ -0,0 +1,333 @@
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.kafka.metadata.migration;
+
+import org.apache.kafka.common.TopicIdPartition;
+import org.apache.kafka.common.Uuid;
+import org.apache.kafka.common.acl.AccessControlEntry;
+import org.apache.kafka.common.config.ConfigResource;
+import org.apache.kafka.common.quota.ClientQuotaEntity;
+import org.apache.kafka.common.resource.ResourcePattern;
+import org.apache.kafka.image.AclsDelta;
+import org.apache.kafka.image.AclsImage;
+import org.apache.kafka.image.ClientQuotasDelta;
+import org.apache.kafka.image.ClientQuotasImage;
+import org.apache.kafka.image.ConfigurationsDelta;
+import org.apache.kafka.image.ConfigurationsImage;
+import org.apache.kafka.image.MetadataDelta;
+import org.apache.kafka.image.MetadataImage;
+import org.apache.kafka.image.TopicImage;
+import org.apache.kafka.image.TopicsDelta;
+import org.apache.kafka.image.TopicsImage;
+import org.apache.kafka.metadata.PartitionRegistration;
+import org.apache.kafka.metadata.authorizer.StandardAcl;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.EnumSet;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.function.BiConsumer;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+public class KRaftMigrationZkWriter {
+    private final MigrationClient migrationClient;
+    private final BiConsumer<String, KRaftMigrationOperation> operationConsumer;
+
+    public KRaftMigrationZkWriter(
+        MigrationClient migrationClient,
+        BiConsumer<String, KRaftMigrationOperation>  operationConsumer
+    ) {
+        this.migrationClient = migrationClient;
+        this.operationConsumer = operationConsumer;
+    }
+
+    public void handleSnapshot(MetadataImage image) {
+        handleTopicsSnapshot(image.topics());
+        handleConfigsSnapshot(image.configs());
+        handleClientQuotasSnapshot(image.clientQuotas());
+        operationConsumer.accept("Setting next producer ID", migrationState ->
+            migrationClient.writeProducerId(image.producerIds().highestSeenProducerId(), migrationState));
+        handleAclsSnapshot(image.acls());
+    }
+
+    public void handleDelta(MetadataImage previousImage, MetadataImage image, MetadataDelta delta) {
```

Comment:
We are missing handling for `clusterDelta` here, right? Specifically, we need to write out KRaft brokers to /brokers. ZK-based brokers will continue to write themselves out.

Reply:
as per the above discussion, this is not needed